wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessTokenFetcher.java

```diff
@@ -49,6 +51,19 @@ PersonalAccessToken fetchPersonalAccessToken(Subject cheUser, String scmServerUr
    * @throws ScmCommunicationException - Some unexpected problem occurred during communication with
    *     scm provider.
    */
+  @Deprecated
   Optional<Boolean> isValid(PersonalAccessToken personalAccessToken)
       throws ScmCommunicationException, ScmUnauthorizedException;
+
+  /**
+   * Checks whether the provided personal access token is valid by fetching user info from the scm
+   * provider. Also checks whether the token has expected scope of permissions if the provider API
+   * supports such request.
+   *
+   * @return - Optional with a pair of boolean value and scm username. The boolean value is true if
+   *     the token has expected scope of permissions, false if the token scopes does not match the
+   *     expected ones. Empty optional if {@link PersonalAccessTokenFetcher} is not able to confirm
+   *     or deny that token is valid.
+   */
+  Optional<Pair<Boolean, String>> isValid(PersonalAccessTokenParams params);
```

Comment:
I'm a little confused by `isValid` being updated to return the username in addition to whether the token is valid. Perhaps these should be separated into multiple functions?

Reply:
The `isValid()` function verifies the token by requesting an scm user. The idea is to return the username in order not to request the username on demand. This reduces the number of API calls.

Reply:
Maybe `getUser()` is a better name?

Reply:
Ah I see the discussion below :)